clientapi/routing/register.go

```diff
@@ -597,6 +625,27 @@ func handleRegistrationFlow(
 		req, r, sessionID, cfg, accountDB, deviceDB)
 }
 
+// Registers immediately a guest account, if guest registration is enabled on the server
+func handleGuestRegistration(
+	req *http.Request,
+	r registerRequest,
+	cfg *config.Dendrite,
+	accountDB *accounts.Database,
+	deviceDB *devices.Database,
+) util.JSONResponse {
+
+	// Exit if guest registration is forbidden
+	if cfg.Matrix.GuestAccessDisabled {
+		return util.MessageResponse(http.StatusForbidden, "Guest access has been disabled")
+	}
+
+	// Create guest user
+	return completeRegistration(
+		req.Context(), accountDB, deviceDB, true, r.Username, r.Password, "",
+		r.InhibitLogin, r.InitialDisplayName, r.DeviceID,
```

Comment:
Won't `r.Username` be at its 0-value (`""`) if we're ignoring everything but `initial_device_display_name` in the JSON body? And even if not, we still need to generate a user ID because `@:server_name` is not a valid MXID. What Synapse does here is use the lowest integer that isn't already in use by another guest user as the localpart, but we could also use a random string (using `util.RandomString`) so that we don't have to scan the whole users table to determine the localpart to use.